src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1002,6 +1002,7 @@ private void StartStream()
                 throw;
             }
 
+            KestrelEventSource.Log.RequestQueued(_currentHeadersStream, AspNetCore.Http.HttpProtocol.Http2);
```

Comment:
While it's just another work item, I think this could be useful to quantify just how backlogged the thread pool is. It's one thing to say there's a thousand pending work items, but another to say there are thousand pending requests.

Reply:
It's literally called `QueueUserWorkitem` 😄 , it's absolutely a queue. We're just counting how may items are ours vs global in the pool like @halter73 says.

Reply:
The same request will end up back in the same threadpool queue as soon as some middleware goes async. What's special about this specific instance? I don't see this producing any actionable data.

Reply:
Have you produced any graphs with this yet? I'm really curious what they'd show.

Reply:
> The same request will end up back in the same threadpool queue as soon as some middleware goes async. 

That's not really true. Actually by default continuations will go into a different queue (well thread local stack). 

> What's special about this specific instance? I don't see this producing any actionable data.

It's useful to see this number be non-zero. While it doesn't capture the entire end to end async flow of the request it can give a sense of things getting held up outside global threadpool queue.

Reply:
> It's useful to see this number be non-zero.

Under sustained load, wouldn't this always be non-zero?

Under what conditions do you expect this number to diverge from it's usual value?

Reply:
> Under sustained load, wouldn't this always be non-zero?

It'll be non-zero if there's blocking or a *huge* amount of load.

> Under what conditions do you expect this number to diverge from it's usual value?

Starvation/blocking and it'll show up under a thing called connections/requests which is why I like it more than just having a single "threadpool queue" counter (though that is useful as well).

Reply:
For a new request on an existing connection, I'm assuming this is the first threadpool dispatch. However. in a threadpool starved scenario, this won't be the first threadpool dispatch for a request on a new connection right?

Am I correct in understanding that only the first scenario (i.e., requests on existing connections) will cause this counter to diverge from zero?

Reply:
For HTTP/1 the request queue is the connection queue for HTTP/2 and 3 we queue per request

Reply:
This got me thinking maybe we should add a PollingCounter for concurrent idle/active connections (e.g. connections (not) in the keep-alive state). It wouldn't be equivalent to `"current-connections" - "current-requests"` because of HTTP/2 and 3. It might even be different for HTTP/1.x if we consider a connection "active" as soon as we get any part of the request line and/or headers.